### PR TITLE
Adds FastPath network support to CNI

### PR DESCRIPTION
**Pull request by trozet**
When a CNI request is made to add a kernel port for a Pod, a second
FastPath port will created as well in a secondary default FastPath
network. This port driver in the future may support multiple high speed
datapath technologies, but for now only supports DPDK. Pods placed in
the same network on their kernel port (original CNI request) will also
be placed in the same FastPath network.

Following the same methodology as userspace CNI, when the FastPath CNI
is invoked, it will annotate the Pod with configuration data and path
information so that the Pod will have access to DPDK port information as
well path to vhost sockets via the downward API. The vhost socket
directory is for now shared amongst all containers and in the future
will be created dynamically by an admission controller. See
k8s/pod_example.yaml for an example of how to deploy a Pod.

Other Changes Include:
 - Fix for an issue where ports were not being deleted after a CNI Del
   operation.
 - Added pod name/ns to protobuf for a future nimbess-CNI to include in
   CNI Request. This avoids having to search for pod name for
   annotation purposes.
 - Changed from govendor to use go modules. This is more consistent with
   Kubernetes and resolves compatibility issues with importing k8s code.

Signed-off-by: Tim Rozet <trozet@redhat.com>

### CODE REVIEW

File: pkg/agent/agent.go [hunk 1]
```diff
@@ -44,21 +51,30 @@ const (
...
 	DriverFailure        = 101
+	FastPathFailure      = 102
 	InternalIPAM         = "nimbess"
+	FastPathNetwork      = "fastpath"
+	DefaultBaseCNIDir    = "/var/lib/nimbess/cni"
+	DefaultPodNs         = "default"
 )
 
-// TODO REMOVE this when IPAM is supported from CNI side
+// For internal Nimbess IPAM
```
Comment (fepan) [inline pkg/agent/agent.go]: keep TODO here. Also, if we are not planning to call external ipam, then also add TODO to change those IPs to ipv6.

File: pkg/agent/agent.go [hunk 2]
```diff
@@ -71,8 +87,12 @@ func getPortName(req *cni.CNIRequest) string {
 }
 
 // invokes IPAM either internal or external and returns the CIDR
-func invokeIPAM(ipamType string) (string, error) {
+func invokeIPAM(ipamType string, fastPath bool) (string, error) {
 	if ipamType == InternalIPAM {
+		if fastPath {
+			ipAddrFast[3]++
+			return fmt.Sprintf("%s/24", ipAddrFast.String()), nil
```
Comment (fepan) [inline pkg/agent/agent.go]: Don't assume ipv4

File: pkg/agent/agent.go [hunk 3]
```diff
@@ -272,9 +297,81 @@ func (s *NimbessAgent) Add(ctx context.Context, req *cni.CNIRequest) (*cni.CNIRe
 		log.Infof("Kernel Virtual Port created: %v in meta network: %s", res.GetInterfaces(), metaKey)
 	}
 
+	// Create FastPath secondary Network
```
Comment (fepan) [inline pkg/agent/agent.go]: remove "secondary"

Comment (trozet) [reply pkg/agent/agent.go]: will remove and make a TODO note to remove creating the network/pipeline automatically. Only the Port will be created in the future unless a UNP applies to this pod.

File: pkg/agent/agent.go (hunk 2 above)
Comment (trozet) [reply pkg/agent/agent.go]: this is just internal IPAM for testing purposes, we really want to use an external IPAM and there is a patch in progress under nimbess-cni to help with this. I'll add a comment.

File: pkg/agent/agent.go (hunk 1 above)
Comment (trozet) [reply pkg/agent/agent.go]: will add

**Pull request merged**